cpp/dolfinx/fem/discreteoperators.h

```diff
@@ -123,4 +123,199 @@ void assemble_discrete_gradient(const fem::FunctionSpace& V0,
+/// a function \f$V_0\f$ into a compatible space \f$V_1\f$, i.e.
+/// \f$V_0 \rightarrow V_1\f$. If \f$u_0\f$ is the
+/// degree-of-freedom vector associated with \f$V_0\f$, then
+/// \f$u_1=Au_0\f$ where \f$u_1\f$ is the degrees-of-freedom vector for
+/// interpolating function in the \f$V_1\f$ space.
+///
+/// @warning This function relies on the user supplying appropriate
+/// input and output spaces. See parameter descriptions.
+///
+/// @param[in] V0 The space to interpolate from
+/// @param[in] V1 The space to interpolate to
+/// @param[in] mat_set A functor that sets values in a matrix
+template <typename T, typename U>
+void assemble_interpolation_matrix(const fem::FunctionSpace& V0,
+                                   const fem::FunctionSpace& V1, U&& mat_set)
+{
+  // Get mesh
+  auto mesh = V0.mesh();
+  assert(mesh);
+
+  // Mesh dims
+  const int tdim = mesh->topology().dim();
+  const int gdim = mesh->geometry().dim();
+
+  // Get elements
+  std::shared_ptr<const FiniteElement> element0 = V0.element();
+  assert(element0);
+  std::shared_ptr<const FiniteElement> element1 = V1.element();
+  assert(element1);
+
+  xtl::span<const std::uint32_t> cell_info;
+  if (element1->needs_dof_transformations()
+      or element0->needs_dof_transformations())
+  {
+    mesh->topology_mutable().create_entity_permutations();
+    cell_info = xtl::span(mesh->topology().get_cell_permutation_info());
+  }
+
+  // Get dofmaps
+  auto dofmap0 = V0.dofmap();
+  auto dofmap1 = V1.dofmap();
+
+  const xt::xtensor<double, 2> X = element1->interpolation_points();
+  // Get block sizes and dof transformation operators
+  const int bs0 = element0->block_size();
+  const int bs1 = element1->block_size();
+  const auto apply_dof_transformation0
+      = element0->get_dof_transformation_function<double>(false, false, false);
+  const auto apply_inverse_dof_transform1
+      = element1->get_dof_transformation_function<T>(true, true, false);
+
+  // Get sizes of elements
+  const std::size_t dim0 = element0->space_dimension() / bs0;
+  const std::size_t value_size_ref0 = element0->reference_value_size() / bs0;
+  const std::size_t value_size0 = element0->value_size() / bs0;
+  const std::size_t value_size1 = element1->value_size() / bs1;
+
+  // Get geometry data
+  const fem::CoordinateElement& cmap = mesh->geometry().cmap();
+  const graph::AdjacencyList<std::int32_t>& x_dofmap
+      = mesh->geometry().dofmap();
+  // FIXME: Add proper interface for num coordinate dofs
+  const std::size_t num_dofs_g = x_dofmap.num_links(0);
+  xtl::span<const double> x_g = mesh->geometry().x();
+
+  // Evaluate coordinate map basis at reference interpolation points
+  xt::xtensor<double, 4> phi(cmap.tabulate_shape(1, X.shape(0)));
+  xt::xtensor<double, 2> dphi;
+  cmap.tabulate(1, X, phi);
+  dphi = xt::view(phi, xt::range(1, tdim + 1), 0, xt::all(), 0);
+
+  // Evaluate v basis functions at reference interpolation points
```

Comment:
Basis functions belong to the function space. Which space? `v` isn't a parameter.